fix($http) No error for json-like repsonses without "application/json" content-type

katsos opened the pull request:
Don't throw error if http response has json-like data but content-type is not "application/json". Instead return data unprocessed.

Fixes https://github.com/angular/angular.js/issues/16027
This is not a breaking change.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)

katsos committed:
fix($http): Don't throw error if http response has json-like data but content-type is not "application/json". Instead return data unprocessed.

fixes #16027

gkalpak commented on src/ng/http.js:
```diff
@@ -141,6 +141,9 @@ function defaultHttpResponseTransform(data, headers) {
         try {
           data = fromJson(tempData);
         } catch (e) {
+          if ((contentType && (contentType.indexOf(APPLICATION_JSON) !== 0))) {
```
Too much parenthesis (even for me :grin:).

gkalpak commented on src/ng/http.js:
```diff
@@ -141,6 +141,9 @@ function defaultHttpResponseTransform(data, headers) {
           data = fromJson(tempData);
         } catch (e) {
+          if ((contentType && (contentType.indexOf(APPLICATION_JSON) !== 0))) {
+            return tempData;
```
Why return `tempData`? If we decide to return the original response, it should be `data`.

gkalpak commented on test/ng/httpSpec.js:
```diff
@@ -1380,8 +1380,28 @@ describe('$http', function() {
+            $http.get('/url').then(callback);
+            $httpBackend.flush();
+
+            expect(callback.calls.mostRecent().args[0].data).toMatch(response);
```
Why not `.toBe()`?

katsos replied:
`toBe()` looks for reference equality, `toMatch()` is responsible for matching strings and regexs. (according to jasmine documentation)

gkalpak replied:
[`toMatch()`][1] is for regular expressions (you can pass a string, but it will be converted to a regular expression). It works in this case, but is unnecessary.

[`toBe()`][2] tests for strict equality (`===`), which works when comparing references to the same object instance or primitives (e.g. strings - as is the case here).

[`toEqual()`][3] can be used for testing loose equality (or with a custom comparator), but it is not necessary here.

[1]: https://github.com/jasmine/jasmine/blob/9cb2f06aa665870dbd5df5fd58a35863738961b3/src/core/matchers/toMatch.js#L14-L28
[2]: https://github.com/jasmine/jasmine/blob/9cb2f06aa665870dbd5df5fd58a35863738961b3/src/core/matchers/toMatch.js#L14-L28
[3]: https://github.com/jasmine/jasmine/blob/9cb2f06aa665870dbd5df5fd58a35863738961b3/src/core/matchers/toEqual.js#L10-L28

gkalpak commented on test/ng/httpSpec.js:
```diff
@@ -1380,8 +1380,28 @@ describe('$http', function() {
+            expect(callback.calls.mostRecent().args[0].data).toMatch(response);
+          });
+
+        });
```
There is no test for not specified `Content-Type` header 😞

katsos committed:
if content-type is not specified return response data unprocessed

katsos committed:
used `toBe()` instead of `toMatch()`
as discussed here https://github.com/angular/angular.js/pull/16075#discussion_r124805676

gkalpak commented on src/ng/http.js:
```diff
@@ -141,6 +141,9 @@ function defaultHttpResponseTransform(data, headers) {
         try {
           data = fromJson(tempData);
         } catch (e) {
+          if (!contentType || contentType && contentType.indexOf(APPLICATION_JSON)) {
```
The `contentType && ` part is unnecessary (since we are already on the right-hand side of the `||`, where `contentType` is truthy. So, the above coondition would become `!contentType || contentType.indexOf(APPLICATION_JSON)`.

While the expression works as intended, it is a little cryptic/hard to read. Since you are essentially trying to negate the already checked `contentType && (contentType.indexOf(APPLICATION_JSON) === 0)`, I suggest you refactor it to something like:

```diff
 if (tempData) {
   var contentType = headers('Content-Type');
-  if ((contentType && (contentType.indexOf(APPLICATION_JSON) === 0)) || isJsonLike(tempData)) {
+  var hasJsonContentType = contentType && (contentType.indexOf(APPLICATION_JSON) === 0);
+  if (hasJsonContentType || isJsonLike(tempData)) {
     try {
       data = fromJson(tempData);
     } catch (e) {
+      if (!hasJsonContentType) {
+        return data;
+      }
       throw $httpMinErr(...);
     }
```

gkalpak commented on test/ng/httpSpec.js:
```diff
@@ -1380,8 +1380,49 @@ describe('$http', function() {
+            var errCallback = jasmine.createSpy('error');
+            $httpBackend.expect('GET', '/url').respond('{abcd}', {'Content-Type': 'text/plain'});
+
+            $http.get('/url').then(callback).catch(errCallback);
```
You don't really need `errCallback` (or the associated assertion) here.

gkalpak commented on test/ng/httpSpec.js:
```diff
@@ -1380,8 +1380,49 @@ describe('$http', function() {
+            var errCallback = jasmine.createSpy('error');
+            $httpBackend.expect('GET', '/url').respond('{abcd}');
+
+            $http.get('/url').then(callback).catch(errCallback);
```
You don't really need `errCallback` (or the associated assertion) here.

gkalpak commented on test/ng/httpSpec.js:
```diff
@@ -1380,8 +1380,49 @@ describe('$http', function() {
+            expect(errCallback).not.toHaveBeenCalled();
+          });
+
+          it('should return response unprocessed if JSON is invalid but content-type is not application/json', function() {
```
The "return response unprocessed" tests could be merged with their "not throw an error"counterparts, but I am fine keeping them as is too.

gkalpak reviewed (commented):
A few more suggestions :grin:

katsos commented:
@gkalpak about failing test, do I have to merge master with my branch?

gkalpak commented:
The failing tests look like flakes. (In any case, they don't seem related to this PR, so don't worry about them for now :smiley:)

katsos committed:
removed unnecessary spies of "errCallback"

katsos committed:
optimized code using `hasJsonContentType` variable - DRY principle

katsos commented:
@gkalpak any progress with this one? thank you for your time!
